chia/wallet/coin_selection.py

```diff
@@ -0,0 +1,62 @@
+import random
+from typing import Set, Optional, Tuple, List
+
+from chia.consensus.default_constants import DEFAULT_CONSTANTS
+from chia.types.blockchain_format.coin import Coin
+from chia.util.ints import uint64
+
+# These algorithms were based off of the algorithms in:
+# https://murch.one/wp-content/uploads/2016/11/erhardt2016coinselection.pdf
+
+
+# we use this to check if one of the coins exactly matches the target.
+def check_for_exact_match(coin_list: List[Coin], target: uint64) -> Optional[Coin]:
+    for coin in coin_list:
+        if coin.amount == target:
+            return coin
+    return None
+
+
+# we use this to find an individual coin greater than the target but as close as possible to the target.
+def find_smallest_coin(greater_coin_list: List[Coin], target: uint64) -> Optional[Coin]:
+    smallest_value = DEFAULT_CONSTANTS.MAX_COIN_AMOUNT  # smallest coins value
+    smallest_coin: Optional[Coin] = None
+    for coin in greater_coin_list:
+        if target < coin.amount < smallest_value:
+            # try to find a coin that is as close as possible to the amount.
+            smallest_value = coin.amount
+            smallest_coin = coin
+    return smallest_coin
+
+
+# we use this to find the set of coins which have total value closest to the target, but at least the target.
+# coins should be sorted in descending order.
+def knapsack_coin_algorithm(smaller_coins: Set[Coin], target: uint64) -> Tuple[Optional[Set[Coin]], uint64]:

```

Comment:
I think it's better to return `Optional[Tuple[Set[Coin], uint64]]`, That is, the uint64 in the optional